Follow BCP 14 rules for SHOULD/SHOULD NOT (#95)
ietf-wg-scitt/draft-ietf-scitt-architecture: draft-ietf-scitt-architecture.md

A DID can be *resolved* into a *key manifest* (a list of public keys indexed by a *key identifier*) using many different DID methods.

> Issuers MAY choose the DID method they prefer, but with no guarantee that all Transparency Services will be able to register their Signed Statements.

Author: TS must advertise the DID methods they support.

Independently, implementations MUST support did:web (but they may be configured not to accept it)

> To facilitate interoperability, all Transparency Service implementations MUST support the `did:web` method {{DID-WEB}}.

Reviewer: One method is "mandatory" to support, other DID Methods are optional, a transparency service MUST support at least 1 did method.

Reviewer 2: For interoperability there are a few choices: either we pick a DID method that is mandatory to implement or we have the TS support all methods listed in the document and clients can select one.

Which approach do you prefer?

Reviewer: @hannestschofenig we discussed wanting a dedicated issue to discuss, and keeping this PR on track.

Reviewer 3: @ad-l  to file a separate issue

Author: https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/98

> To facilitate interoperability, all Transparency Service implementations MUST support the `did:web` method {{DID-WEB}}.

Author: Move this requirement to the TS section

Author: https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/98

> To facilitate interoperability, all Transparency Service implementations MUST support the `did:web` method {{DID-WEB}}.

Reviewer 4: Create a new issue to move this to a new place!

Reviewer: Ping the issue link here to resolve this, thread

For instance, if the Issuer publishes its manifest at `https://sample.issuer/user/alice/did.json`, the DID of the Issuer is `did:web:sample.issuer:user:alice`.

Issuers SHOULD use consistent decentralized identifiers for all their Statements about Artifacts, to simplify authorization by Verifiers and auditing.

> If an issuer uses multiple DIDs (for instance, because their clients support different resolution methods), they MUST ensure that statements signed under each DID are consistent.

Reviewer: The verifier is interested in the authenticity of the issuer of the transparent statement and signed statement... The verifier trusts the TS to vet the issuer of the signed statement... the verifier benefits from the transparency service, if they can't trust the TS, there is not reason to do SCITT.

Reviewer: Tracked in https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/99

> Issuers MAY update their DID manifest at any time, for instance to refresh their signing keys or algorithms, but they SHOULD NOT remove or change any of their previous keys unless they intend to revoke all Signed Statements that are registered as Transparent Statements issued with those keys.

Reviewer: ```suggestion
Issuers MAY update their DID Document at any time, for instance to refresh their signing keys or algorithms, but they SHOULD NOT remove or change any of their previous keys unless they intend to revoke all Signed Statements that are registered as Transparent Statements issued with those keys.
```

> The Issuer's DID appears in the protected header of Signed Statements' Envelopes, while the version of the key from the manifest used to sign the Signed Statement is written in the `kid` header.

Reviewer: ```suggestion
The Issuer's DID appears in the protected header of Signed Statements' Envelopes, while the version of the key from the DID Document used to sign the Signed Statement is written in the `kid` header.
```

Each of these components MUST be carefully protected against both external attacks and internal misbehavior by some or all of the operators of the Transparency Service.
For instance, the code for policy evaluation, Registry extension and endorsement may be protected by running in a TEE; the Registry may be replicated and a consensus algorithm such as Practical Byzantine Fault Tolerance (pBFT {{PBFT}}) may be used to protect against malicious or vulnerable replicas; threshold signatures may be use to protect the service key, etc.

Beyond the trusted components, Transparency Services may operate additional endpoints for auditing, for instance to query for the history of Signed Statements registered by a given Issuer via a certain Feed. Implementations of Transparency Services SHOULD avoid using the service identity and extending the Registry in auditing endpoints, except if it is necessary to compute a Registry consistency proofs. Other evidence to support the correctness and completeness of the audit response MUST be computed from the Registry.

### Service Identity, Remote Attestation, and Keying

> Every Transparency Service MUST have a public service identity, associated with public/private key pairs for signing on behalf of the service.

Reviewer 2: What is a "public service identity"? Let's use terms that are well established. For example, do you want to say that the TS MUST have a certificate and a private key?

> Every Transparency Service MUST have a public service identity, associated with public/private key pairs for signing on behalf of the service.

Reviewer 5: That would be one example. I am not sure, if limiting, for example, to certificate and private signing key is a useful thing to do here. Maybe it is, but it feels rather artificially constraining.

Maybe this is also a symptom of a bigger problem? "Identity" is a hot topic in scitt (as the thread you started shows: https://mailarchive.ietf.org/arch/msg/scitt/kubTlBU7sKLSkTpHdVqudsY7U1Y). Maybe we should dedicate text on that topic (carefully, in order to support to operationalize) based on that discussion and then refer back to a corresponding section?

Other types of cryptographic identities, such as parameters for non-interactive zero-knowledge proof systems, may also be used in the future.

A Transparency Service MAY provide extra evidence that it is securely implemented and operated, enabling remote authentication of the hardware platforms and/or software TCB that run the Transparency Service.
If present, this additional evidence MUST be recorded in the Registry and presented on demand to Verifiers and Auditors.
Examples for Statements that can improve trustworthy assessments of Transparency Services are RATS Conceptual Messages, such as Evidence, Endorsements, or corresponding Attestation Results (see {{-rats-arch}}.

For example, consider a Transparency Service implemented using a set of replicas, each running within its own hardware-protected trusted execution environments (TEEs).
Each replica MAY provide a recent attestation report for its TEE, binding their hardware platform to the software that runs the Transparency Service, the long-term public key of the service, and the key used by the replica for signing Receipts.

> This attestation evidence can be supplemented with transparency Receipts for the software and configuration of the service, as measured in its attestation report.

Reviewer: ```suggestion
This attestation evidence can be supplemented with Receipts for the software and configuration of the service, as measured in its attestation report.
```

Reviewer 2: I don't want to have normative language on attestation in this document because so far attestation hasn't been integrated. It appears strange to suddenly bring it up here.

Reviewer 5: +1

- the ordering of entries, their cryptographic contents, and the Registry governance may be non-deterministic, but they must be verifiable.

> - a Transparency Service MAY store evidence about the resolution of distributed identifiers into manifests.

Reviewer: ```suggestion
- a Transparency Service MAY store evidence about the resolution of DIDs into DID Documents.
```

Reviewer 2: Explain what exactly you want to store.

Receipts are based on COSE Signed Merkle Tree Proofs ({{-COMETRE}}) with an additional wrapper structure that adds the following information:

> - version: Receipt version number; MUST be set to `0` for implementation of this document. We envision that future version of SCITT may add support for more complex receipts; for instance, registrations on multiple TS, receipts for dependency graphs and endorsements of Signed Claims, etc.

Reviewer 2: ```suggestion
- version: Receipt version number; MUST be set to `0` for implementation of this document.
```

Reviewer 2: Everyone understands what the purpose of version numbers are. So, there is no need to talk about what could be done in the future to create a new version

Reviewer 5: +1

> - ts_identifier: The DID of the Transparency Service that issued the claim. Verifiers MAY use this DID as a key discovery mechanism to verify the COSE Merkle Root signature; in this case the verification is the same as for Signed Claims and the signer MAY include the Key ID header. Verifiers MUST support the `did:web` method, all other methods are optional.

Reviewer: I suggest we not invent a new parameter for this, can we use `iss`  via the CWT header draft:
https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/

Reviewer 2: But the TS signs the already signed claim. Wouldn't there be suddently two issuers there?

Reviewer 3: Resolved with: https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/pull/94#discussion_r1330217675

> - ts_identifier: The DID of the Transparency Service that issued the claim. Verifiers MAY use this DID as a key discovery mechanism to verify the COSE Merkle Root signature; in this case the verification is the same as for Signed Claims and the signer MAY include the Key ID header. Verifiers MUST support the `did:web` method, all other methods are optional.

Reviewer: ```suggestion
- ts_identifier: The DID of the Transparency Service that issued the Receipt. Verifiers MAY use this DID as a key discovery mechanism to verify the Receipt; in this case the verification is the same as for Signed Claims and the signer MAY include the `kid` header parameter. Verifiers MUST support the `did:web` method, all other methods are optional.
```

Reviewer 5: +1

- in-toto
- SLSA

> Once the Statement is serialized with the correct media-type/content-format, an Issuer MUST fill in the attributes for the Registration Policy information header.

Reviewer: we should consider "Statement Issuer" vs "Receipt Issuer", for better clarity.

Reviewer 5: Now captured in https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/97

If more than one service is configured, the Verifier MUST return which service the Transparent Statement is registered on.

In some scenarios, the Verifier already expects a specific Issuer and Feed for the Transparent Statement, while in other cases they are not known in advance and can be an output of validation.

> Verifiers MAY be configured to re-verify the Issuer's signature locally, but this requires a fresh resolution of the Issuer's DID, which MAY fail if the manifest is not available or if the statement's signing key has been revoked. Otherwise, the Verifier trusts the validation done by the Transparency Service during Registration.

Reviewer 5: +1
